### DISCUSSION TITLE
[Bug] JavaAgent does not support the default method in the interface

### DISCUSSION THREAD

**Discussion by byspacecbk** [apache/skywalking#9502] [General]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/skywalking/issues?q=is%3Aissue) and found no similar issues.

### Apache SkyWalking Component

Java Agent (apache/skywalking-java)

### What happened

In the springboot project, the controller class implements an interface with a default method (a new feature of Java 8). The default method of this interface can be called, but skywalking does not record it. I have used the latest version of skywalking.

### What you expected to happen

the controller class implements an interface with a default method (a new feature of Java 8)

### How to reproduce

In the springboot project, the controller class implements an interface with a default method (a new feature of Java 8). The default method of this interface can be called, but skywalking does not record it.

public interface ClusterTest2Controller {
    @GetMapping("/_query")
    @QueryAction
    @QueryOperation(summary = "queryPager")
    default Mono<Void> queryPager() {
        System.out.println("=================queryPager=====================");
        return Mono.empty();
    }
}

@RequestMapping("/cluster2")
@RestController
@Authorize
@Tag(name = "cluster2")
public class ClusterTestController implements ClusterTest2Controller{

}

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by wu-sheng** [comment]
You could read the source codes to get the details, https://github.com/apache/skywalking-java/blob/4b564ade6469e9827c63b4e7e5c97df41663000b/apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/AbstractControllerInstrumentation.java#L109

AFAIK, the current plugin requires a class-level annotation(`@RestController` or `@Controller`).

**Reply by byspacecbk** [reply]
Thank you very much for your reply

After trying, I found that there is no problem with inheritance, but not with interface. When I use the interface default method, no matter how I request the default method, skywalking cannot get any endpoint data.

Specific codes are as follows:

1、Use interface implementation，Skywalking cannot capture any data

@RequestMapping("/cluster2")
@RestController
@Authorize
@Tag(name = "cluster2")
public class ClusterTestController **implements** ClusterTest2Controller{
}

@RestController
public **interface** ClusterTest2Controller {
    
    @GetMapping("/_query22")
    @QueryAction
    **default** Mono<Void> queryPager() {
        return Mono.empty();
    }
}

=============================================================

2、Using inheritance, skywalking can capture normal data

@RequestMapping("/cluster2")
@RestController
@Authorize
@Tag(name = "cluster2")
public class ClusterTestController **extends** ClusterTest3Controller{
    
}

@RestController
public class ClusterTest3Controller {

    @GetMapping("/_query33")
    @QueryAction
    public Mono<Void> queryPager2() {
        return Mono.empty();
    }
}

**Reply by byspacecbk** [reply]
After trying, it is found that the method annotated with the interface default method cannot be injected into the spring container, which may be the reason why the interface cannot be correctly captured. Is there any other way to grab this type of interface?

**Reply by wu-sheng** [reply]
> After trying, it is found that the method annotated with the interface default method cannot be injected into the spring container?

What actually does this mean? Why do we need to explain `it can't be injected into spring`? I am confused. I could only guess as following.

Java agent would add a new field to the class, which may not suitable for an interface. This may be the reason. If so, your case is simply not supported.

**Comment by byspacecbk** [comment]
Thank you very much for your reply


After trying, I found that there is no problem with inheritance, but not with interface. When I use the interface default method, no matter how I request the default method, skywalking cannot get any endpoint data.


Specific codes are as follows:


1、Use interface implementation，Skywalking cannot capture any data




=============================================================================================





2、Using inheritance, skywalking can capture normal data









------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/skywalking"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年8月22日(星期一) 下午2:55
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/skywalking] [Bug] JavaAgent does not support the default method in the interface (Discussion #9502)





 
You could read the source codes to get the details, https://github.com/apache/skywalking-java/blob/4b564ade6469e9827c63b4e7e5c97df41663000b/apm-sniffer/apm-sdk-plugin/spring-plugins/mvc-annotation-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/spring/mvc/v4/define/AbstractControllerInstrumentation.java#L109
 
AFAIK, the current plugin requires a class-level ***@***.*** or @Controller).
